raft/confchange/confchange.go

```diff
@@ -0,0 +1,375 @@
+// configuration to arrive at the desired joint configuration.
+func (c Changer) EnterJoint(ccs ...pb.ConfChange) (tracker.Config, tracker.ProgressMap, error) {
+	cfg, prs, err := c.checkAndCopy()
+	if err != nil {
+		return c.err(err)
+	}
+	if joint(cfg) {
+		err := errors.New("config is already joint")
+		return c.err(err)
+	}
+	if len(cfg.Voters[0]) == 0 {
+		err := errors.New("can't make a zero-voter config joint")
+		return c.err(err)
+	}
+	cfg.Voters[1] = quorum.MajorityConfig{}
+	for id := range cfg.Voters[0] {
+		cfg.Voters[1][id] = struct{}{}
+	}
+
+	if err := c.apply(&cfg, prs, ccs...); err != nil {
+		return c.err(err)
+	}
+
+	return cfg, prs, nil
+}
+
+// LeaveJoint transitions out of a joint configuration. It is an error to call
+// this method if the configuration is not joint, i.e. if Voters[1] is empty.
+//
+// The right side of the joint configuration will be removed, and any staged
+// learners (LearnersNext) will be inserted into Learners.
+func (c Changer) LeaveJoint() (tracker.Config, tracker.ProgressMap, error) {
+	cfg, prs, err := c.checkAndCopy()
+	if err != nil {
+		return c.err(err)
+	}
+	if !joint(cfg) {
+		err := errors.New("can't leave a non-joint config")
+		return c.err(err)
+	}
+	if len(cfg.Voters[1]) == 0 {
+		err := fmt.Errorf("configuration is not joint: %v", cfg)
+		return c.err(err)
+	}
+	for id := range cfg.LearnersNext {
+		nilAwareAdd(&cfg.Learners, id)
+		prs[id].IsLearner = true
+	}
+	cfg.LearnersNext = nil
+
+	for id := range cfg.Voters[1] {
+		_, isVoter := cfg.Voters[0][id]
+		_, isLearner := cfg.Learners[id]
+
+		if !isVoter && !isLearner {
+			delete(prs, id)
+		}
+	}
+	cfg.Voters[1] = nil
+	return cfg, prs, nil
+}
+
+// Simple carries out a series of configuration changes that (in aggregate)
+// mutates the set of voters by at most one. It will return an error if that is
+// not the case or if the resulting quorum is zero.
+func (c Changer) Simple(ccs ...pb.ConfChange) (tracker.Config, tracker.ProgressMap, error) {
+	cfg, prs, err := c.checkAndCopy()
+	if err != nil {
+		return c.err(err)
+	}
+	if joint(cfg) {
+		err := errors.New("can't apply simple config change in joint config")
+		return c.err(err)
+	}
+	if err := c.apply(&cfg, prs, ccs...); err != nil {
+		return c.err(err)
+	}
+	if n := symdiff(c.Tracker.Voters[0], cfg.Voters[0]); n > 1 {
+		return tracker.Config{}, nil, errors.New("more than voter changed without entering joint config")
+	}
+	return cfg, prs, nil
+}
+
+func (c Changer) next() uint64 {
+	// We initialize Progress.Next with lastIndex+1 so that the peer will be
+	// probed without an index first.
+	//
+	// TODO(tbg): verify that, this is just my best guess.
+	return c.LastIndex + 1
+}
+
+// apply a ConfChange to the configuration. By convention, changes to voters are
+// always made on the left side (i.e. in Voters[0]), preserving the right hand
+// side as it was when the joint configuration was entered, which is required
+// for correctness.
```

Comment:
> I'd also only use the terms "old" and "new" config (like the dissertation) and get rid of the "left" and "right" notation.

Are you sure this is a good idea? "new config" and "old config" are ambiguous terms. For example, when entering a joint config, the result is a config that you might well think is the new config, but now you really want `Voters[1], Learners` to be the new config. I think we need to find something better here. Some suggestions:

Voters[0] = primary
Voters[1] = secondary / auxiliary

> The practical reason might be that you want cfg.Voters[0] to always correspond to the new config, regardless of whether this is a joint config or not

Yes, that's correct.

I'm going to hold off here until we've had a chance to converge on names.

Reply:
> I also personally don't find "old" and "new" to be particularly ambiguous in this context. I'm basing this mostly off of reading the Raft dissertation, where section 4.3 uses the terms "old", "new", and "transitional" without much confusion.

The Raft thesis is pretty high level, so I don't see the same potential for confusion there neither.   I'm going to go with pre and post as you suggested. (Tomorrow).

Reply:
I gave this a try but I found it unsatisfying and so I tried something else.

pre/post is not great because it's difficult to use in the comments without bending over backwards: 

> EnterJoint verifies that the pre config is empty

It's also annoying that for simple config changes we change the "post" config and the "pre" config does not change (remains empty), this is confusing to me. In the end I think the way things are set up here means that names based on the temporal succession aren't going to work very well. I tried my primary/secondary suggestion and I'm much happier with it:

> // EnterJoint verifies that the secondary (=right) majority config of the joint
// quorum is empty and initializes it with a copy of the primary (=left)
// majority config. That is, it transitions from
//
//     (1 2 3)&&()
// to
//     (1 2 3)&&(1 2 3).

(I don't use left/right elsewhere, but it's helpful for the representation of the config below the comment). I also like that simple config changes just change the "primary" config without populating the "secondary". Similar to old/new it's also clear which way things are moving (we want the primary to be alone).

I worked my way through the comments in this PR using this nomenclature and simplified things a bit as I went, see if you can live with that.

I also held off on introducing the accessors on `JointConfig`, the reason being that they would impose an often awkward duty to use them instead of direct map access, which I found unhelpful. On top of that, `JointConfig` is actually symmetric and the symmetry is only broken by its uses in `confchange`, so I'd preferable introduce the helpers there or in `tracker` if they are ultimately needed, which I'll see in the follow-up PR.

Reply:
I renamed further to `incoming`/`outgoing` which @nvanbenschoten suggested and which I liked.

Reply:
"incoming" and "outgoing" LGTM in context 👍 

> I also held off on introducing the accessors on JointConfig, the reason being that they would impose an often awkward duty to use them instead of direct map access, which I found unhelpful. On top of that, JointConfig is actually symmetric and the symmetry is only broken by its uses in confchange, so I'd preferable introduce the helpers there or in tracker if they are ultimately needed, which I'll see in the follow-up PR.

Helpers in `confchange` would also be fine:
```
func incoming(voters quorum.JointConfig) MajorityConfig { return voters[0] }
func outgoing(voters quorum.JointConfig) MajorityConfig { return voters[1] }
```

The two reasons why I think something like the following would be useful is because:
1. it would serve as in-code documentation for how you're using the `JointConfig` object. Right now, I believe there's only a single comment that describes its use. Other than that, a reader will need to scan the code to figure out how the structure maps to the protocol.
2. it would make the code more easy to read. Compare:
```go
cfg.Voters[1] = quorum.MajorityConfig{}
for id := range cfg.Voters[0] {
    cfg.Voters[1][id] = struct{}{}
}

// to

*outgoingPtr(&cfg.Voters) = quorum.MajorityConfig{}
for id := range incoming(cfg.Voters) {
    outgoing(cfg.Voters)[id] = struct{}{}
}
```

Reply:
Done.